Replace simple activationScripts

nikstur opened the pull request:
This is one part of a series of PRs towards activation without Perl. See more about this larger project here: https://pad.lassul.us/nixos-perlless-activation#

This PR is part of step 1 of the larger project.

In this PR, I replace many of the simple activationScripts. I employed this strategy to remove activationScripts:

1. removing them without replacement because they are not needed anymore
2. replacing them via ExecPreStart=
3. replacing them via a separate service ordered correctly before the other services that need it
4. replacing them via a separate service that runs very early, e.g. with
    ```nix
    wantedBy = [ "sysinit.target" ];
    before = [ "sysinit.target" ];
    unitConfig.DefaultDependencies = false;
    ```

One of the immediate benefits of this work is that these activationScripts now actually run after the initrd when you use the systemd initrd. The systemd initrd calls stage-2-init.sh in the initrd as ./prepare-root

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

nikstur committed:
nixos/nix-daemon: remove activationScript

The activationScript does not seem to be necessary anymore as the paths
are created anyways.

blitz committed:
nixos/timesyncd: replace activationScript via ExecPreStart

nikstur committed:
nixos/activation: replace var activationScript via tmpfiles

nikstur committed:
nixos/activation: remove specialfs activationScript

The stage-2-init.sh script has the same functionality hardcoded so we do
not need it in the activationScript again.

nikstur committed:
nixos/nix-channel: replace activationScript via tmpfiles

nikstur committed:
nixos/binfmt: replace activationScript via tmpfiles

nikstur committed:
nixos/ups: replace activationScript via tmpfiles

nikstur committed:
nixos/mattermost: replace activationScript via tmpfiles

nikstur committed:
nixos/strongswan-swanctl: replace activationScripts via tmpfiles

nikstur committed:
nixos/tests/opensearch: replace activationScript via tmpfiles

nikstur committed:
nixos/tests/grafana: replace activationScript with tmpfiles

nikstur committed:
nixos/stargazer: replace activationScript via tmpfiles

nikstur committed:
nixos/mlmmj: replace activationScript

nikstur committed:
nixos/spiped: replace activationScript via tmpfiles

nikstur committed:
nixos/tests/systemd-timesyncd: replace activationScript via tmpfiles

nikstur committed:
nixos/tests/stunnel: replace activationScript

Replace with a separate systemd service.

nikstur committed:
nixos/activation: link gcroot with tmpfiles instead of in activation

nikstur committed:
nixos/iscsi: replace activationScript

Move the functionality into preStart of the systemd service.

nikstur committed:
nixos/duosec: replace activationScript

Replace with a separate systemd service.

nikstur committed:
nixos/mysql: replace activationScripts via preStart

ElvishJerricco commented on nixos/modules/config/shells-environment.nix:
```diff
@@ -210,14 +210,10 @@ in
-      '';
+    systemd.tmpfiles.rules = [
+      "d /bin 0755 -"
+      "L+ /bin/sh - - - - ${cfg.binsh}"
```
isn't this too late? `/bin/sh` could be used well before `systemd-tmpfiles-setup.service` runs. Same goes for `/usr/bin/env`, though to a lesser degree since `/bin/sh` is used by `system(3)`.

nikstur replied:
I hostely don't know when this is needed. I ran a bunch of tests and they all seemed to be fine. tmpfiles runs before sysinit.target and this is really the earliest we can order anything other than already running it in the initrd.

ElvishJerricco replied:
It's needed by anything that uses `system(3)`, at least. That may not include anything in `systemd`, but who knows what else is out there. I think assuming `/bin/sh` and `/usr/bin/env` is a *bare minimum* that application developers expect of the host system, and those applications may as well be run during early boot.

nikstur replied:
We actually cannot move this into the initrd because this would break initrd-less systems (e.g. containers).  

Normal services are started after basic.target, by which the links already have been established. You would only need /bin/sh earlier if you explicitly order your service earlier. Overall I think its fair to say that the system is not up before basic.target and you shouldn't expect the existence of much. For example you might not even have /tmp or /var/tmp before that, so why expect /bin/sh?

ElvishJerricco replied:
But units ordered before `basic.target` are entirely reasonable, and very well could require `/bin/sh`. The difference between `/bin/sh` and `/tmp` or `/var/tmp` is that `/bin/sh` *has always existed at all times in the system forever*, ***and*** a fundamental system call `system(3)` relies on it.

ElvishJerricco replied:
Of course NixOS has never respected FHS, but there's a reason that an exception was made for `/bin/sh` and `/usr/bin/env`, and I think breaking that would be a bridge too far

ElvishJerricco replied:
> systemd supports hermetic /usr

But you're failing that as well. `/usr/bin/env` in this PR is also ordered far too late.

It's a good point about `/bin/sh` though.

ElvishJerricco commented on nixos/modules/config/users-groups.nix:
```diff
@@ -703,7 +703,7 @@ in {
 
     # Warn about user accounts with deprecated password hashing schemes
-    system.activationScripts.hashes = {
+    system.activationScripts.hashes = if !config.boot.initrd.systemd.enable then {
```
Seems like this warning should make it to somewhere better whether systemd initrd is enabled or not, rather than simply hiding it when systemd initrd is enabled.

nikstur replied:
Not sure if we still need this script at all tbh. Do you have anything in mind where this should go?

ElvishJerricco replied:
Not really :/ There isn't really a place that every user is guaranteed to see an import warning like this.

ElvishJerricco commented on nixos/modules/security/wrappers/default.nix:
```diff
@@ -302,6 +307,11 @@ in
         '';
+      in {
+        Type = "oneshot";
+        ExecStart = "${wrapperScript}";
```
why not just toss the script into the `script` option?

ElvishJerricco commented on nixos/modules/services/hardware/udev.nix:
```diff
@@ -418,18 +418,9 @@ in
-        fi
-      '';
+    systemd.tmpfiles.rules = [
+      "w /sys/module/firmware_class/parameters/path - - - - '${config.hardware.firmware}/lib/firmware'"
```
`systemd-tmpfiles-setup.service` is probably far too late during boot for this. Plenty of devices are found and used arbitrarily early, and they could very well need firmware.

nikstur replied:
Should we throw this into preStart for `systemd-udev` then?

lheckemann replied:
Firmware won't be loaded until the corresponding module is loaded, and the module won't be loaded without udev (or something else in userspace) causing that, so I think that should be fine.

ElvishJerricco replied:
@lheckemann I was under the impression that the kernel itself would use the contents of `/proc/sys/kernel/modprobe` to load modules when devices requiring them are found. Is that false? Because that would mean that both this and the `systemd-modules-load.preStart` have to happen before stage 2 starts. I could be totally wrong though; maybe it's *only* userspace that causes modules to be loaded?

nikstur replied:
Even in the activationScripts no-one guarantees that the kernel might not request modules before the correct script is run. Why does this need to be done before systemd starts?

ElvishJerricco replied:
First of all, `systemd.tmpfiles.rules` is not just after systemd starts; it's *way* after. There are plenty of devices that might need to appear before `local-fs.target`, which are automatically ruled out by having firmware loaded later than that because of `tmpfiles`.

nikstur replied:
So should we order it before `local-fs-pre.target[¶](https://www.freedesktop.org/software/systemd/man/latest/systemd.special.html#local-fs-pre.target)`?

Again I'm very happy to run these as early as you want me  to. I don't have super strong feelings but I need the information when exactly to run them.

ElvishJerricco replied:
I'm also unsure, unfortunately. All I can say is that there's very good reason that this happens before systemd becomes PID 1 currently.

ElvishJerricco commented on nixos/modules/system/boot/modprobe.nix:
```diff
@@ -58,14 +58,9 @@ with lib;
-        echo ${pkgs.kmod}/bin/modprobe > /proc/sys/kernel/modprobe
-      '';
+    systemd.services.systemd-modules-load.preStart = ''
+      echo ${pkgs.kmod}/bin/modprobe > /proc/sys/kernel/modprobe
```
Is this also too late? I seem to recall the kernel being able to load modules itself via the current value of `/proc/sys/kernel/modprobe`, which can happen arbitrarily early I think?

nikstur replied:
If it can really happen arbitrarily early, then this should probably already be done in the initrd. I had a case where a test failed, but setting this as preStart for systemd-modules-load fixed it.

nikstur replied:
This should probably happen before both systemd-modules-load and systemd-udev. These are the only two things that I can think of that would load kernel modules before sysinit.target

lheckemann commented on nixos/modules/system/boot/timesyncd.nix:
```diff
@@ -46,6 +46,29 @@ with lib;
       aliases = [ "dbus-org.freedesktop.timesync1.service" ];
       restartTriggers = [ config.environment.etc."systemd/timesyncd.conf".source ];
+
+      preStart = let
```
`preStart` is to `serviceConfig.ExecStartPre` as `script` is to `serviceConfig.ExecStart`. You don't need an extra writeScript here :)

lheckemann commented on nixos/modules/tasks/network-interfaces.nix:
```diff
@@ -1406,14 +1406,7 @@ in
           val = tempaddrValues.${opt}.sysctl;
          in nameValuePair "net.ipv6.conf.${replaceStrings ["."] ["/"] i.name}.use_tempaddr" val));
 
-    # Set the host and domain names in the activation script.  Don't
```
looks like this landed in the wrong commit

lheckemann commented on nixos/tests/activation/var.nix:
```diff
@@ -0,0 +1,17 @@
+    assert machine.succeed("stat -c '%a' /var/tmp") == "1755\n"
+    assert machine.succeed("stat -c '%a' /var/empty") == "555\n"
+    assert machine.succeed("stat -c '%U' /var/empty") == "root\n"
+    assert machine.succeed("stat -c '%G' /var/empty") == "root\n"
```
Could also check the immutable attr using `lsattr -d`

lheckemann commented on nixos/modules/system/boot/modprobe.nix:
```diff
@@ -58,14 +58,9 @@ with lib;
 
-    system.activationScripts.modprobe = stringAfter ["specialfs"]
-      ''
-        # Allow the kernel to find our wrapped modprobe (which searches
```
The comment is probably worth keeping around.

nikstur committed:
nixos/wrappers: replace activationScript

Create the wrappers via a separate systemd service.

lheckemann commented on nixos/modules/services/monitoring/ups.nix:
```diff
@@ -239,11 +239,9 @@ in
-        mkdir -p /var/state/ups
-      '';
+    systemd.tmpfiles.rules = [
+      "d /var/state/up -"
```
should be `/var/state/ups`

lheckemann commented on nixos/modules/services/networking/strongswan-swanctl/module.nix:
```diff
@@ -43,21 +43,21 @@ in  {
-      mkdir -p '/etc/swanctl/pkcs12'   # PKCS#12 containers
-    '';
+    systemd.tmpfiles.rules = [
+      ''d "/etc/swanctl/x509" -''     # Trusted X.509 end entity certificates
```
I don't think the `-` or the quoting are needed?

lheckemann commented on nixos/modules/services/web-servers/stargazer.nix:
```diff
@@ -204,11 +204,9 @@ in
-        chown -R ${cfg.user}:${cfg.group} /var/lib/gemini/certs
-      '';
+    systemd.tmpfiles.rules = [
+      ''d /var/lib/gemini/certs - "${cfg.user}" "${cfg.group}" -''
```
This has the slight behaviour difference of not changing the ownership of the _contents_. It's probably more correct, and hopefully won't break anything in practice, but maybe @gaykitty could double-check this?

ElvishJerricco replied:
Could add a `Z` line to recursively change the mode and ownership.

gaykitty replied:
The recursive chown should not be necessary as any files created in that directory should already have the correct permissions. This change appears to remove the conditionality of that directory being created, I'd prefer to maintain that if possible. 

I'm not familiar with `systemd.tmpfiles.rules` but it's name and description would suggest it's for *temporary* files, which these cert are not. They should **never** be automatically deleted. Can someone clarify the behavior of this directive for me?

nikstur replied:
Oh right I missed the condition. Readded it. tmpfiles is a poorly chosen name because it can create any file and that's what we are doing here. There is no automatic cleanup.

gaykitty replied:
Sounds good! Then this all seems a okay to me.

ElvishJerricco replied:
@nikstur I'd still recommend a `Z` line unless @gaykitty advises against it. It just makes sure the whole directory is owned by the right  user.

nikstur committed:
nixos/network-interfaces: replace hostname and domain activationScript

The hostname is already set by systemd
https://www.freedesktop.org/software/systemd/man/latest/hostname.html#Hostname%20semantics

Create a separate service that registers the domainname.

lheckemann commented on nixos/modules/tasks/network-interfaces.nix:
```diff
@@ -1406,11 +1406,12 @@ in
+      wantedBy = [ "sysinit.target" ];
+      before = [ "sysinit.target" ];
+      unitConfig.DefaultDependencies = false;
+      script = ''domainname "${cfg.domain}"'';
```
We can avoid the indirection of a shell here by using `serviceConfig.ExecStart`.

l0b0 commented on nixos/modules/security/wrappers/default.nix:
```diff
@@ -275,33 +275,38 @@ in
+
+        # We want to place the tmpdirs for the wrappers to the parent dir.
+        wrapperDir=$(mktemp --directory --tmpdir="${parentWrapperDir}" wrappers.XXXXXXXXXX)
+        chmod a+rx "$wrapperDir"
```
It might be better to use something like `umask 0022` within the previous command (see for example `nixos/tests/stunnel.nix`), then the directory is created with the right permissions immediately.

l0b0 commented on nixos/modules/security/wrappers/default.nix:
```diff
@@ -275,33 +275,38 @@ in
+
+        ${lib.concatStringsSep "\n" mkWrappedPrograms}
+
+        if [ -L ${wrapperDir} ]; then
```
It's [recommended](https://stackoverflow.com/a/669486/96588) to use `[[` instead of `[` in Bash.

l0b0 commented on nixos/modules/security/wrappers/default.nix:
```diff
@@ -275,33 +275,38 @@ in
+          # Atomically replace the symlink
+          # See https://axialcorps.com/2013/07/03/atomically-replacing-files-and-directories/
+          old=$(readlink -f ${wrapperDir})
+          if [ -e "${wrapperDir}-tmp" ]; then
```
Since `"${wrapperDir}-tmp"` is repeated it should probably be a variable.

l0b0 commented on nixos/modules/security/wrappers/default.nix:
```diff
@@ -275,33 +275,38 @@ in
+          rm --force --recursive "$old"
+        else
+          # For initial setup
+          ln --symbolic "$wrapperDir" "${wrapperDir}"
```
I don't think this is robust. This line runs if `$wrapperDir` is not a symlink, but that doesn't necessarily mean that `$wrapperDir` does not exist. If it exists and is a regular file, this command will fail. If it exists and is a directory, the symlink will be created *inside* `$wrapperDir`. Should the condition be `-e` instead of `-L` on line 294?

l0b0 commented on nixos/modules/services/networking/iscsi/initiator.nix:
```diff
@@ -52,25 +52,27 @@ in
+      preStart =
+        let
+          extraCfgDumper = optionalString (cfg.extraConfigFile != null) ''
+            if [ -f "${cfg.extraConfigFile}" ]; then
```
Should this condition use `-s` instead, to check for a non-empty file? Then it would support `cfg.extraConfigFile` being a symlink.

l0b0 commented on nixos/modules/system/boot/timesyncd.nix:
```diff
@@ -46,6 +46,28 @@ with lib;
+          if ! [ -f /var/lib/systemd/timesync/clock ]; then
+            test -d /var/lib/systemd/timesync || mkdir -p /var/lib/systemd/timesync
+            touch /var/lib/systemd/timesync/clock
+          fi
```
This could be simplified as

```bash
mkdir -p /var/lib/systemd/timesync
if ! [[ -f /var/lib/systemd/timesync/clock ]]; then
    touch /var/lib/systemd/timesync/clock
fi
```

`mkdir -p` already checks if the directory exists before attempting to create it, so there's no need to do that separately.

l0b0 reviewed (commented):
I've only reviewed the shell script parts of this PR.

nikstur commented:
> I've only reviewed the shell script parts of this PR.

I will not change the shell scripts themselves. I just move them into a systemd service or preStart.

Improving the shell scripts is out of scope.

lheckemann reviewed (approved):
:heart:

ElvishJerricco reviewed (changes requested):
Some of these orderings are *at least* questionable, if not outright problematic. (leaving this review to negate @lheckemann's premature approval)

nikstur commented:
Resoled @ElvishJerricco comments by removing the changes in question from this PR. I will open another PR where we can have a focused discussion on only the controversial bits.

nikstur commented:
@ofborg test wrappers mysql.mysql80 iscsi-root strongswan-swanctl mattermost systemd-binfmt systemd-timesyncd opensearch.opensearch stunnel grafana.provision activcation-nix-channel activation-var

The stargazer test seems to be broken on master

nikstur commented:
@ElvishJerricco is this good to go now?

ElvishJerricco reviewed (approved):
I think it looks good yea

K900 commented:
https://github.com/NixOS/nixpkgs/commit/a8f50f991948b91d1a191d1348bfff860a39b5a2 broke the installer tests.

K900 commented:
Reverting in https://github.com/NixOS/nixpkgs/pull/264200

nikstur commented:
> Reverting in https://github.com/NixOS/nixpkgs/pull/264200

Thank you for taking the time to only revert the offending commit!

arianvp commented on nixos/modules/tasks/network-interfaces.nix:
```diff
@@ -1406,18 +1406,12 @@ in
-    system.activationScripts.hostname = let
-        effectiveHostname = config.boot.kernel.sysctl."kernel.hostname" or cfg.hostName;
-      in optionalString (effectiveHostname != "") ''
-        hostname "${effectiveHostname}"
```
This is a slight regression in behaviour.  Systemd refuses to set the transient hostname altogether if `/etc/hostname` exists
